vm/src/value.rs

```diff
@@ -84,24 +84,226 @@ impl<'vm, ENV> Iterator for PairIter<'vm, ENV> {
     }
 }
 
-// Do this wrap nonsense so that Value is hashable...
-#[derive(Copy, Clone, Debug)]
-pub struct F32Wrap(pub f32);
-
-impl PartialEq for F32Wrap {
+/**
+ * F56 struct
+ * This struct uses 7 bytes to represent a floating point number.
+ * Most operations are done by converting to f64, performing the operation, and then converting back to F56
+ *
+ * F32 uses 1 bit for the sign,  8 bits for the exponent, and 23 bits for the mantissa.
+ * F56 uses 1 bit for the sign, 10 bits for the exponent, and 45 bits for the mantissa.
+ * F64 uses 1 bit for the sign, 11 bits for the exponent, and 52 bits for the mantissa.
+ *
+ *   Byte 0    Byte 1    Byte 2    Byte 3    Byte 4    Byte 5    Byte 6
+ * [sEEEEEEE][EEEmmmmm][mmmmmmmm][mmmmmmmm][mmmmmmmm][mmmmmmmm][mmmmmmmm]
+ *
+ * Exponent bits range from 0 to 1023
+ * they represent -511 to +512 but are stored biased by +511
+ * the exponent of -511 is reserved for the number 0 and subnormal numbers
+ * the exponent of +512 is reserved for infinity and NaN
+ * so normal exponents range from -510 to +511
+ *
+ * smallest positive subnormal value is 8.48e-168 (2^-555)
+ * smallest positive normal value is 2.98e-154 (2^-510)
+ * maximum finite value is 1.34e154
+*/
+#[derive(Copy, Clone)]
+pub struct F56(pub [u8; 7]);
+impl Eq for F56 {}
+impl PartialEq for F56 {
     fn eq(&self, other: &Self) -> bool {
-        (self.0 - other.0).abs() < f32::EPSILON
-        //self.0.to_bits() == other.0.to_bits()
+        f64::from(*self) == f64::from(*other)
```

Comment:
I think you want to keep the original "< EPSILON" check here otherwise Eq and PartialEq may not be usable.  Implementing these is kinda sloppy at all but I want to be able to hash a Value...

Reply:
Good point. Would it be reasonable to keep F56 acting the same as the primitives and instead add this epsilon diff test elsewhere like in the hash fn? I see some value in making F56 act like the primitives but not if it overcomplicates everything else.

Further, the hash function looks like it converts the f56 to an f64 and then to bits to write a u64. Doesn't this mean that two f56s within EPSILON of each other would still have different bit patterns and hash to different values? I'm not familiar with all the hash so forgive me if I'm going totally the wrong direction with this.

Reply:
Thanks for explaining the conundrum. I now understand why floats are throwing a wrench in the goal to make any Value a proper hash key. There's more I need to understand about the overall project direction in order to make judgment on which imperfect solution we should go forward with. 

In the meantime, I think it would be suitable to implement partial equality for F56 to be 'within EPSILON' and to allow NaN == NaN to be true so we can have the Eq trait. This deviation from primitive float behavior is unfortunate, but the capitalization of F56 in contrast to f64 should help remind everyone that it isn't a primitive. And arguably the behavior is more intuitive than the behavior of primitive floats.

In an effort to merge this PR, I'd like to implement this option and then open a github issue where we can further discuss whether to make F56 and Value implement Eq and Hash, and how to implement PartialEq for F56, and when/whether to panic or throw errors if someone tries to use a float as a hash key.

Reply:
This is reasonable and is a strict improvement over over the old code as well.  Opening an issue is a good idea to hash this out but it is out of scope for this PR.  This is a scripting language so being a bit looser than Rust is probably the right move in most cases (thing Go for things like this).